### DISCUSSION TITLE
guildMemberUpdate & member cache

### DISCUSSION THREAD

**Discussion by DaStormer** [discordjs/discord.js#5585] [Q&A] [answered]
It appears that my nickname/role update logs don't work unless the member is cached beforehand. Does anyone have any insight on how to manage this event w/o cache?

**Comment by kyranet** [comment, marked as answer]
You need the `GUILD_MEMBERS` intent to receive such packets from Discord, as for the uncached members... you can try fetching all members at start-up (although how you do that, is up to you, be careful you can only send 120 packets/minute/shard, and at least 2 are used for heartbeats).

**Reply by DaStormer** [reply]
I do have the GUILD_MEMBERS intent. I'd rather not have all members cached at all times if avoidable since that's very memory intensive. 

I've assumed that the guildMemberUpdate doesn't require the said member to be cached beforehand - is that not correct?

**Reply by kyranet** [reply]
It doesn't if you have enabled the `GUILD_MEMBER` partial (see [Partial](https://discord.js.org/#/docs/main/stable/typedef/PartialType)), but as expected, you won't be able to make a correct comparison if you rely on previous state (to compare roles or nickname).

There's a [feature request](https://github.com/discord/discord-api-docs/issues/2137) in Discord's API docs which would allow us to always emit the difference between the previous and the new state, regardless of our bot's cache.

**Reply by DaStormer** [reply]
Ah, I see. In that case, do you have any suggestions on how to reduce memory usage whilst keeping members cached? Or is high memory usage unavoidable?

**Reply by kyranet** [reply]
The higher the amount of data, the higher the memory usage. It's unavoidable.

Only workaround (which limits the amount of data you can have in the process) is by turning [pointer compression](https://v8.dev/blog/pointer-compression) on, this should be supported in Node.js v14 and newer, although I haven't tried it myself.

**Reply by DaStormer** [reply]
Alright, thanks.
